Build.ps1

```diff
@@ -32,149 +32,145 @@ param(
+
+filter WindowsFilter { param([string]$Version) if($_ -like "*-windowsservercore-$($Version)" -or $_ -like "*-nanoserver-$($windowsVersionMapping[$Version])") { $_ } }
+filter SitecoreFilter { param([string]$Version) if($_ -like "*:$($Version)-windowsservercore-*" -or $_ -like "*:$($Version)-nanoserver-*") { $_ } }
+
+$availableSpecs = Get-BuildSpecifications -Path (Join-Path $PSScriptRoot "\windows")
+
+if(!$IncludeExperimental) {
+    Write-Host "Excluding experimental images."
+    $availableSpecs = $availableSpecs | Where-Object { !$_.Experimental }
+}
+
+$availableTags = $availableSpecs | Select-Object -ExpandProperty Tag
+$defaultTags = $availableTags | Where-Object { $_ -like "mssql-developer:*" -or $_ -like "sitecore-openjdk:*" }
+$miscTags = $availableTags | Where-Object { $_ -like "sitecore-certificates:*" -or $_ -like "sitecore-redis:*" }
+
+$assetTags = $availableTags | Where-Object { $_ -like "sitecore-assets:*" }
+$xmTags = $availableTags | Where-Object { $_ -match "sitecore-xm-(?!sxa|spe|jss).*:.*" }
+$xpTags = $availableTags | Where-Object { $_ -match "sitecore-xp-(?!sxa|spe|jss).*:.*" }
+$xcTags = $availableTags | Where-Object { $_ -match "sitecore-xc-(?!sxa|spe|jss).*:.*" }
+
+$xmSpeTags = $availableTags | Where-Object { $_ -match "sitecore-xm-(spe).*:.*" }
+$xmSxaTags = $availableTags | Where-Object { $_ -match "sitecore-xm-(sxa).*:.*" }
+$xmJssTags = $availableTags | Where-Object { $_ -match "sitecore-xm-(jss).*:.*" }
+
+$xpSpeTags = $availableTags | Where-Object { $_ -match "sitecore-xp-(spe).*:.*" }
+$xpSxaTags = $availableTags | Where-Object { $_ -match "sitecore-xp-(sxa).*:.*" }
+$xpJssTags = $availableTags | Where-Object { $_ -match "sitecore-xp-(jss).*:.*" }
+
+$xcSpeTags = $availableTags | Where-Object { $_ -match "sitecore-xc-(spe).*:.*" }
+$xcSxaTags = $availableTags | Where-Object { $_ -match "sitecore-xc-(sxa).*:.*" }
+
+$knownTags = $defaultTags + $miscTags + $assetTags + $xmTags + $xpTags + $xcTags + $xmSpeTags + $xpSpeTags + $xcSpeTags + $xmSxaTags + $xpSxaTags + $xcSxaTags + $xmJssTags + $xpJssTags
+# These tags are not yet classified and no dependency check is made at this point to know which image it belongs to.
+$catchAllTags = [System.Linq.Enumerable]::Except([string[]]$availableTags, [string[]]$knownTags)
+$catchAllTags | ForEach-Object { $tags.Add($_) > $null }
+
+foreach ($wv in $WindowsVersion) 
 {
-    $tags.Add("mssql-developer:2017-windowsservercore-$($wv)") > $null
-    $tags.Add("sitecore-certificates:latest-nanoserver-$($windowsVersionMapping[$wv])") > $null
-    $tags.Add("sitecore-openjdk:8-nanoserver-$($windowsVersionMapping[$wv])") > $null
-    $tags.Add("sitecore-redis:3.0.504-windowsservercore-$($wv)") > $null
+    $defaultTags | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
+
+    if ($Topology -contains "xp" -or $Topology -contains "xc")
+    {
+        $miscTags | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
```

Comment:
Goes with my above comment, is Redis used in XP?

Reply:
Updated so that xp and xc each have their own list so you don't end up with the extras.